### PR TITLE
Fix #3371: curl_multi_remove_handle() can block until DNS resolution finishes even when using c-ares

### LINKED ISSUE #3371: curl_multi_remove_handle() can block until DNS resolution finishes even when using c-ares
This is related to #2975, but it's not exactly the same issue.

When configured to use c-ares as the DNS resolver, and a handle is still waiting for DNS resolution (such as when the DNS server is unreachable), the `curl_multi_remove_handle()` call can block until that DNS request times out.  

This behaviour appears to have been introduced as part of ac9a179fe980f1a78438a791671e22671ebb6451.

I understand that it might not be possible to cancel an in-progress DNS request with all resolver backends without blocking.  It does seem like it should be possible to cancel an in-progress c-ares DNS request without blocking.

Changing the `multi_done()` call of `Curl_resolver_wait_resolv()` to `Curl_resolver_cancel()` seems to work, but I'm not familiar enough with the details of resolver management inside cURL (especially for other resolver backends) to know if this is correct or "safe" in all cases.  Based on the discussion in #2975, it sounds like the threaded resolver must join with the thread.

Perhaps `Curl_resolver_cancel()` should take a flag that says whether it must not return until all resources are freed.  The c-ares resolver could then simply ignore the flag since has nothing special to do, then clean up and return quickly, while the threaded resolver could block until it has joined with the thread.  

(BTW, it looks like there is at least one case where `Curl_resolver_cancel()` is called when it "shouldn't" wait, in `conn_free()`, but I'm not sure how the threaded resolver's threads get cleaned up in that case.)

### PR DESCRIPTION
**Pull request by b-spencer**
This fixes issue #3371 as discussed and adds a new test case.

New test case 1592 added.  (Sorry if I numbered it wrongly.) 

Tested on Linux x86_64 with c-ares, POSIX threaded, and blocking resolvers.  The threaded resolver behaviour is the same, but now c-ares won't block when handles are removed.

### CODE REVIEW

File: lib/asyn-ares.c [hunk 1]
```diff
@@ -361,11 +372,13 @@ CURLcode Curl_resolver_is_resolved(struct connectdata *conn,
 /*
  * Curl_resolver_wait_resolv()
  *
- * waits for a resolve to finish. This function should be avoided since using
+ * Waits for a resolve to finish. This function should be avoided since using
  * this risk getting the multi interface to "hang".
  *
  * If 'entry' is non-NULL, make it point to the resolved dns entry
  *
+ * This should only be called when 'conn' is in CURLM_STATE_WAITRESOLVE.
```
Comment (b-spencer) [inline lib/asyn-ares.c]: I tried to document the existing calling requirement here.  I didn't attempt to change it.

**Comment by b-spencer** [comment]
It looks like some of the Travis builds failed because of copyright year mismatches.  Is that something I need to do something about?

**Comment by b-spencer** [comment]
I am not sure why https://travis-ci.org/curl/curl/jobs/474913152 is complainint about `ntlm.c` since I didn't touch that file.

I will take a look at the failures in https://travis-ci.org/curl/curl/jobs/474913152

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**